nimsuggest/tests/twithin_macro.nim

```diff
@@ -205,10 +44,21 @@ discard """
 disabled:true
 $nimsuggest --tester $file
 >sug $1
-sug;;skField;;age;;int;;$file;;167;;6;;"";;100;;None
-sug;;skField;;name;;string;;$file;;166;;6;;"";;100;;None
-sug;;skMethod;;twithin_macro.age_human_yrs;;proc (self: Animal): int;;$file;;169;;9;;"";;100;;None
-sug;;skMethod;;twithin_macro.vocalize;;proc (self: Animal): string;;$file;;168;;9;;"";;100;;None
-sug;;skMethod;;twithin_macro.vocalize;;proc (self: Rabbit): string;;$file;;184;;9;;"";;100;;None
+sug;;skField;;age;;int;;$file;;6;;6;;"";;100;;None
+sug;;skField;;name;;string;;$file;;5;;6;;"";;100;;None
+sug;;skMethod;;twithin_macro.age_human_yrs;;proc (self: Animal): int;;$file;;8;;9;;"";;100;;None
+sug;;skMethod;;twithin_macro.vocalize;;proc (self: Animal): string;;$file;;7;;9;;"";;100;;None
+sug;;skMethod;;twithin_macro.vocalize;;proc (self: Rabbit): string;;$file;;23;;9;;"";;100;;None
 sug;;skMacro;;twithin_macro.class;;proc (head: untyped, body: untyped): untyped{.gcsafe, locks: <unknown>.};;$file;;4;;6;;"";;50;;None*
 """
+
+#[
+Notes:
+* here nimsuggest is trying to point out syntax errors rather than hinting
+* this seems to be the first place where things are going off the rails
+* after macros the syntax tree should be transformed for suggestions
+* from the errors that's either broken or not the way it works
+
+Open Questions:
+* how does nimsuggest transform the AST for suggestions or does it even?  
```

Comment:
AFAIK, nimsuggest simply hooks into the compiler `sem` pass, so it does whatever the compiler would do when compiling the program. To get suggestions, nimsuggest have hook points in various places in the `sem` pass, then wait until `sem` process the AST node where the cursor is at, then access the symbol table of that scope to figure out what tokens are available.

Reply:
I had a minor mental breakthrough when I saw the sem pass setup in nimsuggest. The question I wrote was without that nuance, but it's still relevant I think. Even though nimsuggest isn't doing it directly, templates, macros, etc... are all being evaluated, which means lots of computation & state change going on.

At the time I wrote the question and even now, my mind keeps coming back to:
* semantic analysis of any nim code thrown at nimsuggest need to be forgiving
* additionally any compile time execution needs to also be forgiving (templates, macros, ...)
* state (expanded code, computed values, etc...) are going to have interesting impact on analysis and caching

What I'm trying to roughly point at is the above sounds like fertile ground for bugs unless and complexity in terms of usefully (correctly?) carrying out forgiving analysis and if I don't understand those bits better the chance of fixing a test like twithin_macro.nim seem rather slim. :/

Presently, the question could be refined to perhaps, "How does the semantic analysis that nimsuggest carry out via the sem pass transform the AST for generating suggestions, and how does evaluation of compile time code impact things, especially when carrying around or mutating state, carrying out an unbounded computation, employing IO, or allocating memory?". It might lose something in specificity.

IIRC, I've seen bugs like the following:
* nimsuggest goes to 100% and sits there
  * unbounded computation or excessively hitting slow IO path because results weren't cached?
* nimsuggest kept groing memory because of static
  * haven't read the fix yet so not sure what to make of it just yet
* lack of idempotency, I've called the same call twice in a row and I get different results
  * reminds me of Haxe where certain macros must be authored with compilation cache awareness or :(

Reply:
You know the answer to your questions already. There is some logic that prevents IO from being done in the VM for nimsuggest but all the other problems are there. However, ignoring the fact that plenty of code wasn't written with these restrictions in mind (hence bugs), the restrictions are not that hard to follow: There can be error types and nodes in the AST and processing should ignore them and continue.

Reply:
@Araq thanks for confirming -- also provides some hints for potentially quick PRs.

I'm reading through notes and open questions and cleaning them up accordingly.